src/highdicom/legacy/sop.py

```diff
@@ -25,385 +37,391 @@
-                continue
-            elif tag in mf_attributes:
-                mf_dataset.add(da)
+        """
+        self._frames = single_frame_list
+        self._distinguishing_attributes_tags = distinguishing_tags
+        tmp = [
+            tag_for_keyword('AcquisitionDateTime'),
+            tag_for_keyword('AcquisitionDate'),
+            tag_for_keyword('AcquisitionTime'),
+            tag_for_keyword('SpecificCharacterSet')]
+        self._excluded_from_perframe_tags =\
+            self._distinguishing_attributes_tags + tmp
+        self._perframe_tags: List[BaseTag] = []
+        self._shared_tags: List[BaseTag] = []
+        self._find_per_frame_and_shared_tags()
+
+    @property
+    def frames(self) -> List[Dataset]:
+        return self._frames[:]
+
+    @property
+    def distinguishing_attributes_tags(self) -> List[Tag]:
+        return self._distinguishing_attributes_tags[:]
+
+    @property
+    def excluded_from_perframe_tags(self) -> List[Tag]:
+        return self._excluded_from_perframe_tags[:]
+
+    @property
+    def perframe_tags(self) -> List[Tag]:
+        return self._perframe_tags[:]
+
+    @property
+    def shared_tags(self) -> List[Tag]:
+        return self._shared_tags[:]
+
+    @property
+    def series_instance_uid(self) -> UID:
+        """Returns the series instance uid of the _FrameSet"""
+        return self._frames[0].SeriesInstanceUID
+
+    @property
+    def study_instance_uid(self) -> UID:
+        """Returns the study instance uid of the _FrameSet"""
+        return self._frames[0].StudyInstanceUID
+
+    def get_sop_instance_uid_list(self) -> list:
+        """Returns a list containing all SOPInstanceUID of the _FrameSet"""
+        output_list = [f.SOPInstanceUID for f in self._frames]
+        return output_list
+
+    def get_sop_class_uid(self) -> UID:
+        """Returns the sop class uid of the _FrameSet"""
+        return self._frames[0].SOPClassUID
+
+    def _find_per_frame_and_shared_tags(self) -> None:
+        """Detects and collects all shared and perframe attributes"""
+        rough_shared: dict = {}
+        sfs = self.frames
+        for ds in sfs:
+            for ttag, elem in ds.items():
+                if (not ttag.is_private and not
+                    _DicomHelper.istag_file_meta_information_group(ttag) and not
+                        _DicomHelper.istag_repeating_group(ttag) and not
+                        _DicomHelper.istag_group_length(ttag) and not
+                        self._istag_excluded_from_perframe(ttag) and
+                        ttag != tag_for_keyword('PixelData')):
+                    elem = ds[ttag]
+                    if ttag not in self._perframe_tags:
+                        self._perframe_tags.append(ttag)
+                    if ttag in rough_shared:
+                        rough_shared[ttag].append(elem.value)
+                    else:
+                        rough_shared[ttag] = [elem.value]
```

Comment:
I don't believe using fancy data containers would make the code more readable, since this awkward checking has long been the common way of handling similar use cases in python and not difficult to follow. But anyway, I will go ahead and apply what you suggest.